### PR TITLE
Redecomp sparse matrix transfers

### PR DESCRIPTION
**Pull request by ebchin**
Implements sparse matrix transfers using redecomp.  Currently supports L2 spaces of order 0 and vdim = 1 (1 value per element).

### CODE REVIEW

**Comment by jamiebramwell** [comment]
I assume this is to help solve @kswartz92 's filter problem?

**Comment by ebchin** [comment]
> I assume this is to help solve @kswartz92 's filter problem?

That's correct

**Comment by ebchin** [comment]
@jamiebramwell and @srwopschall this one is ready for review. Thanks!

File: src/redecomp/transfer/SparseMatrixTransfer.cpp [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+{
+  SLIC_ERROR_ROOT_IF(static_cast<const mfem::Mesh*>(&redecomp_test_mesh_.getParent()) != parent_test_space_.GetMesh(),
+    "The parent test quadrature space mesh must be linked to the test Redecomp mesh.");
+  SLIC_ERROR_ROOT_IF(static_cast<const mfem::Mesh*>(&redecomp_trial_mesh_.getParent()) != parent_trial_space_.GetMesh(),
+    "The parent trial quadrature space mesh must be linked to the trial Redecomp mesh.");
+  SLIC_ERROR_ROOT_IF(&redecomp_test_mesh_.getMPIUtility().MPIComm() != &redecomp_trial_mesh_.getMPIUtility().MPIComm(),
+    "MPI Communicator must match in test and trial spaces.");
+
+  // one value per element for initial implementation
+  SLIC_ERROR_ROOT_IF(!parent_test_space_.IsDGSpace() || !parent_trial_space_.IsDGSpace() ||
```
Comment (jamiebramwell) [inline src/redecomp/transfer/SparseMatrixTransfer.cpp:40]: @kswartz92 , is there a need to make this more general than L2 order 0 in the near term?

File: src/redecomp/transfer/SparseMatrixTransfer.cpp [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+    // test space = row space
+    auto test_parent_rank = 0;
+    auto test_ghost_ct = 0;
+    for (int i{0}; i < src.Height(); ++i)
+    {
+      const auto test_elem_id = i; // assumes 1 point per element
+      while (test_elem_id == test_elem_offsets[test_parent_rank + 1])
+      {
+        ++test_parent_rank;
+        test_ghost_ct = 0;
```
Comment (jamiebramwell) [inline src/redecomp/transfer/SparseMatrixTransfer.cpp:115]: Why do you have to keep setting this to zero?

File: src/redecomp/transfer/SparseMatrixTransfer.cpp [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+    auto src_J = src.GetJ();
+    auto src_data = src.GetData();
+
+    auto& test_elem_offsets = redecomp_test_mesh_.getRedecompToParentElemOffsets();
+    auto& test_ghost_elems = redecomp_test_mesh_.getRedecompToParentGhostElems();
+    auto& trial_elem_offsets = redecomp_trial_mesh_.getRedecompToParentElemOffsets();
+
+    // test space = row space
+    auto test_parent_rank = 0;
+    auto test_ghost_ct = 0;
```
Comment (jamiebramwell) [inline src/redecomp/transfer/SparseMatrixTransfer.cpp:107]: What is the `test_ghost_ct`?

File: src/redecomp/transfer/SparseMatrixTransfer.cpp [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+      // skip unowned elements
+      if (test_ghost_ct < test_ghost_elems[test_parent_rank].size() 
+        && test_elem_id == test_ghost_elems[test_parent_rank][test_ghost_ct])
+      {
+        ++test_ghost_ct;
+        continue;
+      }
+      // trial space = col space
+      auto trial_parent_rank = 0;
+      for (int Ij{src_I[i]}; Ij < src_I[i+1]; ++Ij)
```
Comment (jamiebramwell) [inline src/redecomp/transfer/SparseMatrixTransfer.cpp:129]: I think some words explaining this algorithm would be welcome.

**Review by jamiebramwell — APPROVED** [review]

**Comment by ebchin** [comment]
Thanks for taking a look @jamiebramwell, I believe I've addressed your comments in the latest commit.  @srwopschall, do you mind taking a quick look before merging?

File: src/redecomp/transfer/SparseMatrixTransfer.cpp (hunk 1 above)
Comment (kswartz92) [reply src/redecomp/transfer/SparseMatrixTransfer.cpp:40]: I don't think so, L2 order 0 will cover most use cases.

File: src/redecomp/transfer/SparseMatrixTransfer.cpp [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+  }
+  // create hypre par matrix
+  return std::make_unique<mfem::HypreParMatrix>(getMPIUtility().MPIComm(),
+    parent_test_space_.GlobalVSize(), parent_trial_space_.GlobalVSize(), 
+    parent_test_space_.GetDofOffsets(), parent_trial_space_.GetDofOffsets(),
+    i_diag, j_diag, data_diag, i_offd, j_offd, data_offd, cmap_j_offd.size(), cmap
+  );
+}
+
+const RedecompMesh& SparseMatrixTransfer::getRedecompMesh(const mfem::FiniteElementSpace& quadrature_space)
```
Comment (srwopschall) [inline src/redecomp/transfer/SparseMatrixTransfer.cpp]: rename input argument. Not a quadrature space.

File: src/redecomp/RedecompMesh.cpp [hunk 6]
```diff
@@ -125,13 +127,18 @@ RedecompMesh::RedecompMesh(
...
 EntityIndexByRank RedecompMesh::BuildP2RElementList(
   const Partitioner& partitioner,
-  int n_parts
+  int n_parts,
+  double ghost_length
 ) const
 {
+  if (ghost_length < 0.0)
+  {
+    ghost_length = 1.25 * MaxElementSize(parent_, partitioner.getMPIUtility());
```
Comment (srwopschall) [inline src/redecomp/RedecompMesh.cpp]: Add comment as to what this is calculating or the rationale behind this choice.

File: src/tests/redecomp_sparsematrix.cpp [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+};
+
+TEST_P(SparseMatrixTest, mass_matrix_transfer)
+{
+  EXPECT_LT(max_error_, 1.0e-13);
+
+  MPI_Barrier(MPI_COMM_WORLD);
+}
+
+INSTANTIATE_TEST_SUITE_P(redecomp, SparseMatrixTest, testing::Values(
```
Comment (srwopschall) [inline src/tests/redecomp_sparsematrix.cpp:186]: Let's double check test runtime in CI. Let's double check that each refinement level is in fact testing something different. Perhaps refinement 1 and 2 are not testing anything new, but 3 guarantees zero element redecomp meshes, etc.

File: src/examples/element_matrix_redecomp.cpp [hunk 8]
```diff
@@ -4,7 +4,7 @@
 // SPDX-License-Identifier: (MIT)
 
 /**
- * @file matrix_redecomp.cpp
+ * @file element_matrix_redecomp.cpp
  *
  * @brief Demonstrates moving a matrix using redecomp::MatrixTransfer
```
Comment (srwopschall) [inline src/examples/element_matrix_redecomp.cpp]: Consider "Demonstrates moving element matrices using redecomp::MatrixTransfer"; this may be a little more precise since now we have multiple definitions of what "a matrix" is.

File: src/examples/sparse_matrix_redecomp.cpp [hunk 9]
```diff
@@ -0,0 +1,242 @@
+// Copyright (c) 2017-2023, Lawrence Livermore National Security, LLC and
+// other Tribol Project Developers. See the top-level LICENSE file for details.
+//
+// SPDX-License-Identifier: (MIT)
+
+/**
+ * @file sparse_matrix_redecomp.cpp
+ *
+ * @brief Demonstrates moving a matrix using redecomp::SparseMatrixTransfer
```
Comment (srwopschall) [inline src/examples/sparse_matrix_redecomp.cpp]: Consider "Demonstrates moving a sparse matrix using redecomp::SparseMatrixTransfer"; it's a little pedantic, but more precise since we have multiple matrix transfers now.

File: src/examples/sparse_matrix_redecomp.cpp [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+ *
+ * Demonstrates use of redecomp::SparseMatrixTransfer to move a symmetric mfem::SparseMatrix whose rows and columns are
+ * associated with a mfem::FiniteElementSpace on a redecomp::RedecompMesh to its representation as an
+ * mfem::HypreParMatrix on the associated mfem::ParMesh and mfem::ParFiniteElementSpace. The row and column finite
+ * element space are the same in this example, but redecomp::SparseMatrixTransfer supports rectangular matrices with
```
Comment (srwopschall) [inline src/examples/sparse_matrix_redecomp.cpp]: spaces

File: src/examples/sparse_matrix_redecomp.cpp [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+ * solution for the transformed field.
+ *
+ * Example runs (from repo root directory):
+ *   - mpirun -np 4 {build_dir}/examples/sparse_matrix_redecomp_ex -d 2 -r 0 -R 0.1
+ *   - mpirun -np 4 {build_dir}/examples/sparse_matrix_redecomp_ex -d 2 -r 3 -R 0.4
+ *   - mpirun -np 4 {build_dir}/examples/sparse_matrix_redecomp_ex -d 3 -r 0 -R 0.4
+ *   - mpirun -np 4 {build_dir}/examples/sparse_matrix_redecomp_ex -d 3 -r 2 -R 0.1
+ *
+ * @details This example relies on using a cartesian mesh with uniform element size to obtain an analytical definition
+ * of the filtered field. Do not change the mesh to anything other than a square without adjusting the analytical
```
Comment (srwopschall) [inline src/examples/sparse_matrix_redecomp.cpp]: square/cube?

File: src/redecomp/RedecompMesh.cpp [hunk 12]
```diff
@@ -125,13 +127,18 @@ RedecompMesh::RedecompMesh(
 
 EntityIndexByRank RedecompMesh::BuildP2RElementList(
   const Partitioner& partitioner,
-  int n_parts
+  int n_parts,
+  double ghost_length
 ) const
 {
+  if (ghost_length < 0.0)
```
Comment (srwopschall) [inline src/redecomp/RedecompMesh.cpp]: Out of curiosity, what would happen if ghost length was 0.0? Or if ghost length was some small positive number that wasn't the 'length' of adjacent ghost elements? Maybe we guard for this possibility?

File: src/redecomp/RedecompMesh.hpp [hunk 13]
```diff
@@ -148,9 +150,14 @@ class RedecompMesh : public mfem::Mesh
    * 
    * @param partitioner Method of partitioning the elements
    * @param n_parts Number of parts to partition the mesh into
+   * @param ghost_length Length around the domain border to include ghost elements
```
Comment (srwopschall) [inline src/redecomp/RedecompMesh.hpp]: Minor, but it is really more of a "fuzz" width beyond the current domain border. Length around the domain border makes it sound like you only include ghost elements on a portion of the domain border.

File: src/redecomp/transfer/SparseMatrixTransfer.cpp [hunk 14]
```diff
@@ -0,0 +1,321 @@
...
+)
+: parent_test_space_ { parent_test_space },
+  parent_trial_space_ { parent_trial_space },
+  redecomp_test_space_ { redecomp_test_space },
+  redecomp_trial_space_ { redecomp_trial_space },
+  redecomp_test_mesh_ { getRedecompMesh(redecomp_test_space_) },
+  redecomp_trial_mesh_ { getRedecompMesh(redecomp_trial_space_) }
+{
+  SLIC_ERROR_ROOT_IF(static_cast<const mfem::Mesh*>(&redecomp_test_mesh_.getParent()) != parent_test_space_.GetMesh(),
+    "The parent test quadrature space mesh must be linked to the test Redecomp mesh.");
```
Comment (srwopschall) [inline src/redecomp/transfer/SparseMatrixTransfer.cpp]: Change quadrature space to finite element space?

File: src/redecomp/transfer/SparseMatrixTransfer.cpp [hunk 15]
```diff
@@ -0,0 +1,321 @@
...
+  SLIC_ERROR_ROOT_IF(static_cast<const mfem::Mesh*>(&redecomp_trial_mesh_.getParent()) != parent_trial_space_.GetMesh(),
+    "The parent trial quadrature space mesh must be linked to the trial Redecomp mesh.");
```
Comment (srwopschall) [inline src/redecomp/transfer/SparseMatrixTransfer.cpp]: Quadrature to finite element

File: src/redecomp/transfer/SparseMatrixTransfer.cpp [hunk 16]
```diff
@@ -0,0 +1,321 @@
...
+      cmap_val.second = offd_ct++;
+    }
+  }
+  // create cmap, offd column map from local to global element IDs
+  mfem::Memory<HYPRE_BigInt> cmap(cmap_j_offd.size());
+  for (auto& cmap_val : cmap_j_offd)
+  {
+    cmap[cmap_val.second] = cmap_val.first;
+  }
+  // compute j_diag and j_offd, the columns for each row (row offsets given by i).  will not be sorted.
```
Comment (srwopschall) [inline src/redecomp/transfer/SparseMatrixTransfer.cpp]: populate vs. compute? Maybe elaborate on "will not be sorted".

File: src/redecomp/transfer/SparseMatrixTransfer.cpp (hunk 4 above)
Comment (srwopschall) [reply src/redecomp/transfer/SparseMatrixTransfer.cpp:129]: @ebchin - Per @jamiebramwell 's comment, this is a complex algorithm that relies on an inherent understanding or familiarity with CSR arrays, some underlying knowledge of the structure of a hypre par matrix, and the mpi communication. Per her comment, please review for any extra explanation that may help elaborate on this algorithm.

File: src/redecomp/transfer/SparseMatrixTransfer.cpp [hunk 17]
```diff
@@ -0,0 +1,321 @@
...
+  auto offd_nnz = 0;
+  // NOTE: these will be owned by the HypreParMatrix. using mfem::Memory, the data will not be deleted upon destruction.
+  mfem::Memory<HYPRE_Int> i_diag(parent_test_space_.GetVSize() + 1);
+  mfem::Memory<HYPRE_Int> i_offd(parent_test_space_.GetVSize() + 1);
+  for (int i{0}; i < parent_test_space_.GetVSize() + 1; ++i)
+  {
+    i_diag[i] = 0;
+    i_offd[i] = 0;
+  }
+  // maps from global j to order of appearance in the loops below
```
Comment (srwopschall) [inline src/redecomp/transfer/SparseMatrixTransfer.cpp]: Consider clarity of comment. "in the loops below" isn't super clear. Global j may be hard to parse for someone not immediately refreshed on CSR data. This is complex code to be sure, but I'm not sure that this comment enhances understanding of the next handful of lines.

File: src/redecomp/transfer/SparseMatrixTransfer.cpp [hunk 18]
```diff
@@ -0,0 +1,321 @@
...
+      }
+      else
+      {
+        ++offd_nnz;
+        ++i_offd[recv_parent_local_row[r][i] + 1];
+        cmap_j_offd.insert(std::make_pair(col_starts[col_rank] + recv_parent_local_col[r][i], cmap_j_offd.size()));
+      }
+    }
+  }
+  // sum up i_diag and i_offd (completing their definition)
```
Comment (srwopschall) [inline src/redecomp/transfer/SparseMatrixTransfer.cpp]: Consider clarity of comment. completing their definition is a little unclear.

File: src/redecomp/transfer/SparseMatrixTransfer.cpp [hunk 19]
```diff
@@ -0,0 +1,321 @@
...
+    parent_test_space_.GlobalVSize(), parent_trial_space_.GlobalVSize(), 
+    parent_test_space_.GetDofOffsets(), parent_trial_space_.GetDofOffsets(),
+    i_diag, j_diag, data_diag, i_offd, j_offd, data_offd, cmap_j_offd.size(), cmap
+  );
+}
+
+const RedecompMesh& SparseMatrixTransfer::getRedecompMesh(const mfem::FiniteElementSpace& quadrature_space)
+{
+  auto redecomp_mesh = dynamic_cast<const RedecompMesh*>(quadrature_space.GetMesh());
+  SLIC_ERROR_ROOT_IF(redecomp_mesh == nullptr, "The quadrature space must have a Redecomp mesh.");
```
Comment (srwopschall) [inline src/redecomp/transfer/SparseMatrixTransfer.cpp]: Rename quadrature space.

File: src/redecomp/transfer/SparseMatrixTransfer.hpp [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+   * 
+   * @param fe_space Finite element space linked to a RedecompMesh
+   * @return const RedecompMesh& Redecomp mesh associated with finite element space
+   */
+  static const RedecompMesh& getRedecompMesh(const mfem::FiniteElementSpace& fe_space);
+
+  /**
+   * @brief Build vector of parent element offsets for each MPI rank
+   * 
+   * @param redecomp_mesh RedecompMesh whose parent linked mesh the element offset vector is desired
```
Comment (srwopschall) [inline src/redecomp/transfer/SparseMatrixTransfer.hpp]: Wording is confusing.

File: src/redecomp/transfer/SparseMatrixTransfer.hpp [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+   */
+  const mfem::FiniteElementSpace& redecomp_test_space_;
+
+  /**
+   * @brief Trial finite element space on the RedecompMesh
+   */
+  const mfem::FiniteElementSpace& redecomp_trial_space_;
+
+  /**
+   * @brief Test RedecompMesh
```
Comment (srwopschall) [inline src/redecomp/transfer/SparseMatrixTransfer.hpp:108]: Not strictly required, but consider a note somewhere indicating that since you support rectangular sparse matrices you need to have two redecomp meshes. Up to now the paradigm has only been one on-rank redecomp mesh for a given redecomposition. It wouldn't hurt elaborating on why, even if internally, two are required.

File: src/tests/redecomp_sparsematrix.cpp [hunk 22]
```diff
@@ -0,0 +1,223 @@
...
+#include "mfem.hpp"
+
+#include "tribol/config.hpp"
+
+#include "redecomp/redecomp.hpp"
+
+namespace redecomp {
+
+/**
+ * @brief This test transfers an mfem::SparseMatrix whose rows and colums are associated with an order-0 L2 field on a
```
Comment (srwopschall) [inline src/tests/redecomp_sparsematrix.cpp]: Mentione if this is symmetric like the example.

File: src/tests/redecomp_sparsematrix.cpp [hunk 23]
```diff
@@ -0,0 +1,223 @@
...
+        double Wij = filter_kernel(xi, xj);
+        // if kernel returns a non-zero value, create entry in the sparse redecomp matrix
+        if (Wij > 0.0) {
+          W_redecomp.Add(i, j, Wij);
+          n_row_entries[i]++;
+        }
+      }
+    }
+
+    W_redecomp.Finalize();
```
Comment (srwopschall) [inline src/tests/redecomp_sparsematrix.cpp:104]: Consider noting that this is required in order to interface with the redecomp capability.

**Review by srwopschall — APPROVED** [review]
@ebchin - This looks excellent. I've made minor clarity comments that require addressing.

File: src/redecomp/RedecompMesh.cpp (hunk 12 above)
Comment (ebchin) [reply src/redecomp/RedecompMesh.cpp]: This should be fine -- there would be no ghost elements if the ghost length is 0.0.  Created a few new constructors so the ghost length based on the max element size isn't based on the value of the ghost length passed in.

**Comment by ebchin** [comment]
Thanks for the review @srwopschall.  I've addressed your comments in the latest commits.
